src/lib/grammar_ast.rs

```diff
@@ -0,0 +1,171 @@
+use std::collections::{HashMap, HashSet};
+use std::fmt;
+
+pub struct GrammarAST {
```

Comment:
But what does AST stand for? Abstract Syntax Tree? Because I'm not sure there is a tree involved.